modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpPipeliningHandler.java

```diff
@@ -227,8 +276,28 @@ private boolean doFlush(ChannelHandlerContext ctx) {
         }
         boolean needsFlush = true;
         while (channel.isWritable()) {
-            final WriteOperation currentWrite = queuedWrites.poll();
+            WriteOperation currentWrite = queuedWrites.poll();
             if (currentWrite == null) {
+                drainQueuedNowWritable(ctx);
+                if (channel.isWritable() == false) {
+                    // try flushing to make channel writable again, loop will only continue if channel becomes writable again
+                    ctx.flush();
+                    needsFlush = false;
+                }
+                if (channel.isWritable() == false) {
+                    break;
+                }
+                currentWrite = queuedWrites.poll();
+            }
+            if (currentWrite == null) {
+                // no write from a non-chunked message was found queued to be written, check if a chunked message might have become
+                // writable
+                if (currentChunkedWrite != null) {
+                    if (writeChunk(ctx, currentChunkedWrite.combiner, currentChunkedWrite.response.body())) {
+                        finishChunkedWrite();
+                        continue;
```

Comment:
I'd think we would want to continue rather than break also if `writeChunk` returns false?

Reply:
Right :facepalm: that was a real bug, thanks for catching. I added a specific test with a message that outputs small chunks to reproduce and cover now (see https://github.com/elastic/elasticsearch/pull/88311/commits/9b0680009263bfb991833bc5c30b2d9ec6489293).